Individual part weight support for balanced partitioning

N-Maas opened the pull request:
To avoid any boredom, let's open the next pull request :smile:

It includes two major changes:
- Support for individual part weights in the balanced partitioning algorithm introduced in #70. Now, the according parameters are compatible with individual part weights and the same balance guarantees apply (ok, I have no formal proof for this, but I'm relatively sure the original proof can be adapted to individual part weights).
- The way the balance constraint is handled for individual part weights is changed (independent on whether the hypergraph is weighted or not). Currently, each bisection is allowed to use the maximum feasible weight for each part in the case of individual part weights, leading to more restrictions for lower level bisections and the refinement. With this PR, the adaptive epsilon approach is transferred to individual part weights by reversely calculating an epsilon from the given maximum part weights at each bisection. This should improve quality for individual part weights even for unweighted hypergraphs.

N-Maas commented on kahypar/partition/bin_packing/bin_packing_utils.h:
```diff
@@ -208,68 +245,5 @@ static inline void preassignFixedVertices(const Hypergraph& hg, const std::vecto
-  }
-
-  return max;
-}
```
This functionality is moved to bin_packer.h (currentBinImbalance/partitionIsDeeplyBalanced) in a slightly more general fashion: It is now applicable to individual part weights and it is abstracted over different load balancing algorithms (thus the need to call it with dynamic dispatch).

SebastianSchlag commented:
> To avoid any boredom, let's open the next pull request 😄
> 

Woohoo! 🙌 

> It includes two major changes:
> 
> * Support for individual part weights in the balanced partitioning algorithm introduced in #70. Now, the according parameters are compatible with individual part weights and the same balance guarantees apply (ok, I have no formal proof for this, but I'm relatively sure the original proof can be adapted to individual part weights).
> * The way the balance constraint is handled for individual part weights is changed (independent on whether the hypergraph is weighted or not). Currently, each bisection is allowed to use the maximum feasible weight for each part in the case of individual part weights, leading to more restrictions for lower level bisections and the refinement. With this PR, the adaptive epsilon approach is transferred to individual part weights by reversely calculating an epsilon from the given maximum part weights at each bisection. This should improve quality for individual part weights even for unweighted hypergraphs.

That sounds like a good idea. 👍  Are you running some experiments for this?

N-Maas commented:
> That sounds like a good idea. :+1: Are you running some experiments for this?

At the moment it is only tested manually (and in the integration test), but a small experiment might make sense. I wonder what is the best was to design it? Probably we want use identical part weights per instance (e.g. just using the weights the usual balance constraint yields), as otherwise we need a way to determine varying part weights without complicating the experimental setup too much, and of course it should be deterministic.

SebastianSchlag commented:
> > That sounds like a good idea. 👍 Are you running some experiments for this?
> 
> At the moment it is only tested manually (and in the integration test), but a small experiment might make sense. I wonder what is the best was to design it? Probably we want use identical part weights per instance (e.g. just using the weights the usual balance constraint yields), as otherwise we need a way to determine varying part weights without complicating the experimental setup too much, and of course it should be deterministic.

Following up on this, @N-Maas did you run the some experiments on a small set of unweighted instances to make sure that everything still works as expected?

SebastianSchlag commented:
This instance posted here: https://github.com/kahypar/kahypar/issues/84 might also be a good test case (after adjusting the vertex weights) - especially in combination with fixed vertices.

If I add 1 to each of the vertex weights and run 

`./kahypar/application/KaHyPar  --hypergraph ~/Downloads/tmp1.hgr.txt -k 2 -e 0.5 -o cut -m direct -p ~/Downloads/cut.ini.txt --verbose=1 --part-weights 100 94 --use-individual-part-weights=true -f ~/Downloads/tmp1.fix.txt`

on master, I get the following error message:

> WARNING: Fixed vertex subgraph is imbalanced. Switching to direct k-way initial partitioning!
> [bin_packing_utils.h:setPart:46]: Assertion `bin >= 0 && index < _parts.size()` failed: Invalid bin id:  bin = 1
> [1]    5493 abort      ./kahypar/application/KaHyPar --hypergraph ~/Downloads/tmp1.hgr.txt -k 2 -e

Would that work on your branch?

N-Maas commented:
> > > That sounds like a good idea. +1 Are you running some experiments for this?
> > 
> > 
> > At the moment it is only tested manually (and in the integration test), but a small experiment might make sense. I wonder what is the best was to design it? Probably we want use identical part weights per instance (e.g. just using the weights the usual balance constraint yields), as otherwise we need a way to determine varying part weights without complicating the experimental setup too much, and of course it should be deterministic.
> 
> Following up on this, @N-Maas did you run the some experiments on a small set of unweighted instances to make sure that everything still works as expected?

No, not yet. But I will try to find some time for it soon.

N-Maas commented:
> This instance posted here: #84 might also be a good test case (after adjusting the vertex weights) - especially in combination with fixed vertices.
> 
> If I add 1 to each of the vertex weights and run
> 
> `./kahypar/application/KaHyPar --hypergraph ~/Downloads/tmp1.hgr.txt -k 2 -e 0.5 -o cut -m direct -p ~/Downloads/cut.ini.txt --verbose=1 --part-weights 100 94 --use-individual-part-weights=true -f ~/Downloads/tmp1.fix.txt`
> 
> on master, I get the following error message:
> 
> > WARNING: Fixed vertex subgraph is imbalanced. Switching to direct k-way initial partitioning!
> > [bin_packing_utils.h:setPart:46]: Assertion `bin >= 0 && index < _parts.size()` failed: Invalid bin id:  bin = 1
> > [1]    5493 abort      ./kahypar/application/KaHyPar --hypergraph ~/Downloads/tmp1.hgr.txt -k 2 -e
> 
> Would that work on your branch?

I looked into it and interestingly, this error has no connection to individual part weights or the prepacking algorithm at all. It happens in the bin packing initial partitioner, but is caused by the use of fixed vertices:
In initial_partition::partition, we switch to direct k-way initial partitioning because the fixed vertex subgraph is imbalanced (initial_partition.h, line 162). However, here the initial context instead of a recursive bisection context is used, which is not properly initialized for initial partitioning (at least for the assumptions of the bin packing initial partitioner). Specifically, `context.partition.rb_upper_k` and `context.partition.rb_lower_k` are not set, which causes the error.

So on my branch, the behavior is exactly the same.

I wonder whether it might be best to remove this check entirely (or only print a warning without changing the behavior of the algorithm), as now there is no benefit in using direct k-way initial partitioning anyways (the balance won't be better then that of the prepacking algorithm). This would also remove all associated edge cases.

N-Maas commented:
> > > That sounds like a good idea. +1 Are you running some experiments for this?
> > 
> > 
> > At the moment it is only tested manually (and in the integration test), but a small experiment might make sense. I wonder what is the best was to design it? Probably we want use identical part weights per instance (e.g. just using the weights the usual balance constraint yields), as otherwise we need a way to determine varying part weights without complicating the experimental setup too much, and of course it should be deterministic.
> 
> Following up on this, @N-Maas did you run the some experiments on a small set of unweighted instances to make sure that everything still works as expected?

I have run some experiments on benchmark set D, running master and my branch both with and without individual part weights (direct k-way mode). The results without individual part weights are exactly the same as on master. With individual part weights activated (by setting 1+e times average block weight manually for each part) the performance has improved compared to master, but is still a bit worse than without individual part weights.

Also I have fixed the bug with the instance from #84 and removed the special case for fixed vertices in initial_partition (as discussed with @kittobi1992).

SebastianSchlag commented:
> > > > That sounds like a good idea. +1 Are you running some experiments for this?
> > > 
> > > 
> > > At the moment it is only tested manually (and in the integration test), but a small experiment might make sense. I wonder what is the best was to design it? Probably we want use identical part weights per instance (e.g. just using the weights the usual balance constraint yields), as otherwise we need a way to determine varying part weights without complicating the experimental setup too much, and of course it should be deterministic.
> > 
> > 
> > Following up on this, @N-Maas did you run the some experiments on a small set of unweighted instances to make sure that everything still works as expected?
> 
> I have run some experiments on benchmark set D, running master and my branch both with and without individual part weights (direct k-way mode). The results without individual part weights are exactly the same as on master. With individual part weights activated (by setting 1+e times average block weight manually for each part) the performance has improved compared to master, but is still a bit worse than without individual part weights.
> 
> Also I have fixed the bug with the instance from #84 and removed the special case for fixed vertices in initial_partition (as discussed with @kittobi1992).

Sounds good! Thanks, @N-Maas! 👍
